src/Internal/BalanceTx/BalanceTx.purs

```diff
@@ -1,936 +1,92 @@
-      , transaction: unbalancedCollTx
-      , changeAddress
-      , changeDatum: changeDatum'
-      , allUtxos
-      , utxos: availableUtxos
-      , miscFee: getCertsBalance transaction pparams + getProposalsBalance
-          transaction
-      }
-  where
-  getChangeAddress :: BalanceTxM Address
-  getChangeAddress = maybe (liftContract Wallet.getChangeAddress) pure
-    =<< asksConstraints Constraints._changeAddress
-
-  transactionWithNetworkId :: BalanceTxM Transaction
-  transactionWithNetworkId = do
-    networkId <- maybe askNetworkId pure
-      (transaction ^. _body <<< _networkId)
-    pure (transaction # _body <<< _networkId ?~ networkId)
-
-setTransactionCollateral :: Address -> Transaction -> BalanceTxM Transaction
-setTransactionCollateral changeAddr transaction = do
-  nonSpendableSet <- asksConstraints _nonSpendableInputs
-  mbCollateralUtxos <- asksConstraints _collateralUtxos
-  -- We must filter out UTxOs that are set as non-spendable in the balancer
-  -- constraints
-  let isSpendable = not <<< flip Set.member nonSpendableSet
-  collateral <- case mbCollateralUtxos of
-    -- if no collateral utxos are specified, use the wallet, but filter
-    -- the unspendable ones
-    Nothing -> do
-      let isSpendableUtxo = isSpendable <<< _.input <<< unwrap
-      { yes: spendableUtxos, no: filteredUtxos } <-
-        Array.partition isSpendableUtxo <$> do
-          liftEitherContract $ note CouldNotGetCollateral <$>
-            Wallet.getWalletCollateral
-      when (not $ Array.null filteredUtxos) do
-        logWarn' $ pprintTagSet
-          "Some of the collateral UTxOs returned by the wallet were marked as non-spendable and ignored"
-          (pprintUtxoMap (TransactionUnspentOutputs.toUtxoMap filteredUtxos))
-      pure spendableUtxos
-    -- otherwise, get all the utxos, filter out unspendable, and select
-    -- collateral using internal algo, that is also used in KeyWallet
-    Just utxoMap -> do
-      ProtocolParameters params <- liftContract getProtocolParameters
-      let
-        maxCollateralInputs = UInt.toInt $ params.maxCollateralInputs
-        mbCollateral =
-          Array.fromFoldable <$>
-            selectCollateral params.coinsPerUtxoByte maxCollateralInputs utxoMap
-      liftEither $ note (InsufficientCollateralUtxos utxoMap) mbCollateral
-  addTxCollateralReturn collateral (addTxCollateral collateral transaction)
-    changeAddr
-
---------------------------------------------------------------------------------
--- Balancing Algorithm
---------------------------------------------------------------------------------
-
-type BalancerParams =
-  { strategy :: SelectionStrategy
-  , transaction :: Transaction
-  , changeAddress :: Address
-  , changeDatum :: Maybe OutputDatum
-  , allUtxos :: UtxoMap
-  , utxos :: UtxoMap
-  , miscFee :: BigInt -- can be negative (deregistration)
+type CtlBalancerContext =
+  { balancerConstraints :: BalancerConstraints
+  , extraUtxos :: UtxoMap
```

Comment:
What's the meaning of `extraUtxos` field? I would prefer some documentation comment with this field.

I'd expect it means additional UTXOs, however, that has a different meaning and is set via the `balancerConstraints`. So I need further explanation for this... :thinking: